### PR TITLE
raftstore: ignore async fetch result when the peer is pending removed

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/11973, Close https://github.com/tikv/tikv/issues/12026

What's Changed:
When the peer has been already removed, it cleans the raft log totally. After that, if there is an async fetch result, it results in calling `term()` in the handler of `raft-rs`, and cause panic due to raft log is not found which is deleted just before.

To solve that, ignore the async fetch result if the peer has been removed.

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -1186,23 +1187,34 @@ where
...
-                self.fsm.has_ready = true;
+                self.on_raft_log_fetched(context, res);
             }
         }
     }
 
+    fn on_raft_log_fetched(&mut self, context: GetEntriesContext, res: Box<RaftlogFetchResult>) {
+        let low = res.low;
+        // if the peer is being destroyed, ignore the result.
+        if self.fsm.peer.pending_remove {
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:1198]: Seems it's a subset of https://github.com/tikv/tikv/issues/12026. Can you also fix #12026 in this PR?

File: tests/failpoints/cases/test_async_fetch.rs [hunk 2]
```diff
@@ -103,3 +102,69 @@ fn test_node_cache_compact_with_one_node_down() {
     }
     panic!("cluster is not compacted after inserting 200 entries.");
 }
+
+// Test the case where async fetch is performed while the peer is removed.
```
Comment (BusyJay) [inline tests/failpoints/cases/test_async_fetch.rs:106]:
```suggestion
// Test the case that async fetch is performed while the peer is removed.
```

File: tests/failpoints/cases/test_async_fetch.rs [hunk 3]
```diff
@@ -103,3 +102,69 @@ fn test_node_cache_compact_with_one_node_down() {
...
+    sleep_ms(10);
+    fail::remove("pause_on_peer_collect_message");
+
+    cluster.pd_client.must_add_peer(1, new_peer(1, 1));
+
+    // logs should be replicated to node 1 successfully.
+    for i in 1..60 {
+        let k = i.to_string().into_bytes();
+        let v = k.clone();
+        must_get_equal(cluster.engines[&1].kv.as_inner(), &k, &v);
```
Comment (BusyJay) [inline tests/failpoints/cases/test_async_fetch.rs:168]: You mean `must_get_equal(&cluster.engines[&1].kv, &k, &v);`?

File: tests/failpoints/cases/test_async_fetch.rs [hunk 4]
```diff
@@ -103,3 +102,69 @@ fn test_node_cache_compact_with_one_node_down() {
...
+    cluster.cfg.raft_store.raft_log_reserve_max_ticks = 2;
+    cluster.cfg.raft_store.raft_entry_cache_life_time = ReadableDuration::millis(100);
+    cluster.run();
+
+    cluster.must_put(b"k1", b"v1");
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+
+    let mut before_states = HashMap::default();
+
+    for (&id, engines) in &cluster.engines {
```
Comment (BusyJay) [inline tests/failpoints/cases/test_async_fetch.rs:126]: The check seems meaningless.

**Review by NingLin-P — APPROVED** [review]

File: components/raftstore/src/store/fsm/peer.rs [hunk 5]
```diff
@@ -1186,23 +1187,34 @@ where
...
-                self.fsm.has_ready = true;
+                self.on_raft_log_fetched(context, res);
             }
         }
     }
 
+    fn on_raft_log_fetched(&mut self, context: GetEntriesContext, res: Box<RaftlogFetchResult>) {
+        let low = res.low;
+        // if the peer is not the leader anymore or being destroyed, ignore the result.
+        if !self.fsm.peer.is_leader() || self.fsm.peer.pending_remove {
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:1198]: The test case doesn't cover if `is_leader` is true while `pending_remove` is false.

Comment (Connor1996) [reply components/raftstore/src/store/fsm/peer.rs:1198]: add a case to reproduce the panic, PTAL

File: tests/failpoints/cases/test_async_fetch.rs [hunk 6]
```diff
@@ -103,3 +102,122 @@ fn test_node_cache_compact_with_one_node_down() {
...
+        let v = k.clone();
+        must_get_equal(&cluster.get_engine(1), &k, &v);
+    }
+}
+
+// Test the case that async fetch is performed well after the leader has stepped
+// down to follower and is applying snapshot.
+#[test]
+fn test_node_async_fetch_leader_change() {
+    test_util::init_log_for_test();
```
Comment (BusyJay) [inline tests/failpoints/cases/test_async_fetch.rs:159]: Should be deleted.

**Review by BusyJay — COMMENTED** [review]
The cases can pass without the patch.

**Comment by Connor1996** [comment]
> The cases can pass without the patch.

interesting, I can't reproduce either. Seems I make something wrong when doing the final clean.

**Comment by Connor1996** [comment]
> The cases can pass without the patch.

Oh, because we update raft-rs which avoids panic internally. To reproduce, you should revert raft-rs update as well.

the panic would be like this
```
async_fetch::test_node_async_fetch_leader_change stdout ----
thread 'raftstore-1-0::test_node_async_fetch_leader_change' panicked at '[region 1] 1 unexpected state: Applying(0)', /data4/zbk/tikv/components/raftstore/src/store/peer_storage.rs:1283:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/core/src/panicking.rs:143:14
   2: raftstore::store::peer_storage::PeerStorage<EK,ER>::snapshot
             at /data4/zbk/tikv/components/raftstore/src/store/peer_storage.rs:1283:13
   3: <raftstore::store::peer_storage::PeerStorage<EK,ER> as raft::storage::Storage>::snapshot
             at /data4/zbk/tikv/components/raftstore/src/store/peer_storage.rs:736:9
   4: raft::raft_log::RaftLog<T>::snapshot
             at /data4/zbk/.cargo/git/checkouts/raft-rs-42b8049ef2e3af07/2a3d7b6/src/raft_log.rs:468:9
   5: raft::raft::RaftCore<T>::prepare_send_snapshot
             at /data4/zbk/.cargo/git/checkouts/raft-rs-42b8049ef2e3af07/2a3d7b6/src/raft.rs:670:26
   6: raft::raft::RaftCore<T>::maybe_send_append
             at /data4/zbk/.cargo/git/checkouts/raft-rs-42b8049ef2e3af07/2a3d7b6/src/raft.rs:827:25
   7: raft::raft::RaftCore<T>::send_append
             at /data4/zbk/.cargo/git/checkouts/raft-rs-42b8049ef2e3af07/2a3d7b6/src/raft.rs:760:9
   8: raft::raft::Raft<T>::send_append
             at /data4/zbk/.cargo/git/checkouts/raft-rs-42b8049ef2e3af07/2a3d7b6/src/raft.rs:877:9
   9: raft::raw_node::RawNode<T>::on_entries_fetched
             at /data4/zbk/.cargo/git/checkouts/raft-rs-42b8049ef2e3af07/2a3d7b6/src/raw_node.rs:430:21
  10: raftstore::store::fsm::peer::PeerFsmDelegate<EK,ER,T>::on_raft_log_fetched
             at /data4/zbk/tikv/components/raftstore/src/store/fsm/peer.rs:1278:9
  11: raftstore::store::fsm::peer::PeerFsmDelegate<EK,ER,T>::on_significant_msg
             at /data4/zbk/tikv/components/raftstore/src/store/fsm/peer.rs:1255:17
  12: raftstore::store::fsm::peer::PeerFsmDelegate<EK,ER,T>::handle_msgs
             at /data4/zbk/tikv/components/raftstore/src/store/fsm/peer.rs:651:49
  13: <raftstore::store::fsm::store::RaftPoller<EK,ER,T> as batch_system::batch::PollHandler<raftstore::store::fsm::peer::PeerFsm<EK,ER>,raftstore::store::fsm::store::StoreFsm<EK>>>::handle_normal
             at /data4/zbk/tikv/components/raftstore/src/store/fsm/store.rs:789:9
  14: batch_system::batch::Poller<N,C,Handler>::poll
             at /data4/zbk/tikv/components/batch-system/src/batch.rs:406:27
  15: batch_system::batch::BatchSystem<N,C>::start_poller::{{closure}}
             at /data4/zbk/tikv/components/batch-system/src/batch.rs:572:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'cases::test_async_fetch::test_node_async_fetch_leader_change' panicked at 'can't get value Some("100") for key 313030', components/test_raftstore/src/util.rs:75:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/core/src/panicking.rs:143:14
   2: test_raftstore::util::must_get
             at /data4/zbk/tikv/components/test_raftstore/src/util.rs:75:5
   3: test_raftstore::util::must_get_equal
             at /data4/zbk/tikv/components/test_raftstore/src/util.rs:83:5
   4: failpoints::cases::test_async_fetch::test_node_async_fetch_leader_change
             at ./failpoints/cases/test_async_fetch.rs:221:9
   5: failpoints::cases::test_async_fetch::test_node_async_fetch_leader_change::{{closure}}
             at ./failpoints/cases/test_async_fetch.rs:158:1
   6: core::ops::function::FnOnce::call_once
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/core/src/ops/function.rs:227:5
   7: test_util::runner::run_test_with_hook::{{closure}}::{{closure}}
             at /data4/zbk/tikv/components/test_util/src/runner.rs:64:21
   8: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/core/src/ops/function.rs:227:5
   9: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/alloc/src/boxed.rs:1854:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

File: tests/failpoints/cases/test_async_fetch.rs [hunk 7]
```diff
@@ -103,3 +102,121 @@ fn test_node_cache_compact_with_one_node_down() {
...
+
+    // make sure destroy_peer and on_entries_fetched are called in one batch.
+    cluster.must_transfer_leader(1, new_peer(2, 2));
+    fail::cfg("worker_gc_raft_log", "pause").unwrap();
+    cluster.pd_client.must_remove_peer(1, new_peer(1, 1));
+    fail::cfg("pause_on_peer_collect_message", "pause").unwrap();
+    fail::remove("worker_gc_raft_log");
+    sleep_ms(10);
+    fail::remove("worker_async_fetch_raft_log");
+    sleep_ms(10);
```
Comment (BusyJay) [inline tests/failpoints/cases/test_async_fetch.rs:142]: Every subtle.

**Review by BusyJay — APPROVED** [review]

File: tests/failpoints/cases/test_async_fetch.rs (hunk 7 above)
Comment (Connor1996) [reply tests/failpoints/cases/test_async_fetch.rs:142]: make it be prone to follow the order that destroy peer is handled before on entries fetched

**Comment by BusyJay** [comment]
Seems these test_delete_lock_proposed_after_proposing_locks{,_1,_2} fail frequently. /cc @sticnarf 

/run-test

**Pull request merged**